### PR TITLE
fix(#1555): incorrect exe highlight in Windows filesystem from WSL

### LINKED ISSUE #1555: WSL: Files from mounted Windows filesystem have incorrect colouring
### Description

Files in tree have incorrect red colour after opening a mounted directory to Windows filesystem from WSL.

### Neovim version

```text
NVIM v0.7.2
Build type: RelWithDebInfo
LuaJIT 2.1.0-beta3
```

### Operating system and version

Windows 10, WSL Ubuntu 20.04

### nvim-tree version

c3ea264947671f44d836af5b7587e12c4b4611f9

### Minimal config

```lua
Template nvt-min.lua is enough to reproduce this issue
Note: Issue occurs even with git disabled
```

### Steps to reproduce

1. Setup and run WSL instance, then perform all next steps from within WSL
2. Clone some Git repository to Windows filesystem (e.g. `/mnt/c/Users/doanamo/Desktop/repo`)
3. Navigate to cloned repository and run `nvim -nu /tmp/nvt-min.lua`
4. Observe the issue

### Expected behavior

Issue does not occur if repository is cloned to Linux filesystem from WSL.

image

### Actual behavior

image

This colouring makes it hard to identify file statuses from Git.

### PR DESCRIPTION
**Pull request by doanamo**
Resolves #1555

### CODE REVIEW

File: lua/nvim-tree/explorer/node-builders.lua [hunk 1]
```diff
@@ -4,6 +4,10 @@ local watch = require "nvim-tree.explorer.watch"
 
 local M = {
   is_windows = vim.fn.has "win32" == 1,
+  is_wsl = (function()
+    local proc_version = io.open("/proc/version"):read("*a")
```
Comment (alex-courtis) [inline lua/nvim-tree/explorer/node-builders.lua:8]: Blows up when not found. Set to `io.open("/procccccccccc/version")`.

Need to test the results of io.open.

```
Error executing vim.schedule lua callback: ...vt-dev/site/pack/packer/start/packer.nvim/lua/packer.lua:339: Vim(lua):E5108: Error executing lua ...ree.lua.doanamo/lua/nvim-tree/explorer/node-builders.lua:8: attempt to index a nil value
stack traceback:
        ...ree.lua.doanamo/lua/nvim-tree/explorer/node-builders.lua:8: in function <...ree.lua.doanamo/lua/nvim-tree/explorer/node-builders.lua:7>
        ...ree.lua.doanamo/lua/nvim-tree/explorer/node-builders.lua:10: in main chunk
        [C]: in function 'require'
        ...nvim-tree.lua.doanamo/lua/nvim-tree/explorer/explore.lua:4: in main chunk
        [C]: in function 'require'
        ...rt/nvim-tree.lua.doanamo/lua/nvim-tree/explorer/init.lua:9: in main chunk
        [C]: in function 'require'
        ...acker/start/nvim-tree.lua.doanamo/lua/nvim-tree/core.lua:2: in main chunk
        [C]: in function 'require'
        ...rt/nvim-tree.lua.doanamo/lua/nvim-tree/renderer/init.lua:1: in main chunk
        [C]: in function 'require'
        ...packer/start/nvim-tree.lua.doanamo/lua/nvim-tree/lib.lua:3: in main chunk
        [C]: in function 'require'
        ...ack/packer/start/nvim-tree.lua.doanamo/lua/nvim-tree.lua:4: in main chunk
        [C]: in function 'require'
        /tmp/nvt-dev.lua:32: in function 'setup'
        [string ":lua"]:1: in main chunk
        [C]: in function 'cmd'
        ...vt-dev/site/pack/packer/start/packer.nvim/lua/packer.lua:339: in function ''
        vim/_editor.lua: in function <vim/_editor.lua:0>
stack traceback:
        [C]: in function 'cmd'
        ...vt-dev/site/pack/packer/start/packer.nvim/lua/packer.lua:339: in function ''
        vim/_editor.lua: in function <vim/_editor.lua:0>
```

**Review by alex-courtis — CHANGES_REQUESTED** [review]
- [ ] test io.open result

Reviewed but not tested: no available windows environment.

All codepaths but `is_wsl` guarded by windows checks and will be accepted without verification.

**Comment by alex-courtis** [comment]
`:help has` indicates that "wsl" is a possibility; can you use that?

File: lua/nvim-tree/explorer/node-builders.lua (hunk 1 above)
Comment (alex-courtis) [inline lua/nvim-tree/explorer/node-builders.lua:8]: This handle must be closed.

**Comment by doanamo** [comment]
Updated with following changes:
* Use `has "wsl"` as suggested
* Close `io.popen` handles
* Greatly improved performance by reducing the amount of `wslpath` calls using lazy evaluation and making calls for parents of file nodes (directories/links) instead of for each file
  * Tested performance by navigating directories such as `/usr/bin/` and `C:\Windows\System32`
* Fix reported code style

Thank you for your review!

**Comment by doanamo** [comment]
Updated again:
* I've forgot to update a call to `is_executable()` after changing its signature which resulted in errors on reloading tree view
* I wasn't properly calling `close()` on handles

File: lua/nvim-tree/explorer/node-builders.lua [hunk 2]
```diff
@@ -4,6 +4,7 @@ local watch = require "nvim-tree.explorer.watch"
 
 local M = {
   is_windows = vim.fn.has "win32" == 1,
+  is_wsl = vim.fn.has "wsl" == 1,
```
Comment (alex-courtis) [inline lua/nvim-tree/explorer/node-builders.lua:7]: :+1:

File: lua/nvim-tree/explorer/node-builders.lua [hunk 3]
```diff
@@ -24,9 +25,19 @@ function M.folder(parent, absolute_path, name)
   }
 end
 
-function M.is_executable(absolute_path, ext)
+function M.is_executable(parent, absolute_path, ext)
   if M.is_windows then
     return utils.is_windows_exe(ext)
+  elseif M.is_wsl then
+    if parent.is_wsl_windows_fs_path == nil then
+      -- Evaluate lazily when needed and do so only once for each parent
```
Comment (alex-courtis) [inline lua/nvim-tree/explorer/node-builders.lua:33]: :+1:

File: lua/nvim-tree/utils.lua [hunk 4]
```diff
@@ -182,6 +186,36 @@ function M.is_windows_exe(ext)
   return M.pathexts[ext:upper()]
 end
 
+function M.is_wsl_windows_fs_path(path)
```
Comment (alex-courtis) [inline lua/nvim-tree/utils.lua:189]: Please add doc to both these methods as per `is_windows_exe`

**Review by alex-courtis — CHANGES_REQUESTED** [review]
Everything is guarded behind `is_wsl` and `is_windows`.

I cannot test; assuming that this works. Please be available to resolve any issues that users report.

- [ ] [luadoc](https://keplerproject.github.io/luadoc/manual.html) for new utils

**Comment by alex-courtis** [comment]
tip: you can run the linter locally: [CONTRIBUTING.md](https://github.com/kyazdani42/nvim-tree.lua/blob/master/CONTRIBUTING.md)

git-hooks are optional but useful

**Comment by doanamo** [comment]
Updated with luadoc comments and style fixes

**Comment by doanamo** [comment]
> Please be available to resolve any issues that users report.

Sure thing! Feel free to ping me in any case that may seem related to my changes. 👍

**Comment by alex-courtis** [comment]
> > Please be available to resolve any issues that users report.
> 
> Sure thing! Feel free to ping me in any case that may seem related to my changes. +1

Thank you. If there's any issue we will mention you.

**Review by alex-courtis — APPROVED** [review]

**Pull request merged**